suite/tests/client-interface/drreg-test.c

```diff
@@ -765,6 +831,181 @@ GLOBAL_LABEL(FUNCNAME:)
+        ud2
+        /* xax is dead, so should not need to spill when reserving aflags. */
+        mov      REG_XAX, 0
+        jmp      epilog15
+     epilog15:
+        add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
+        POP_CALLEE_SAVED_REGS()
+        ret
+        /* This test does not have AArchXX variants. */
+#elif defined(ARM)
+        bx       lr
+#elif defined(AARCH64)
+        ret
+#endif
+        END_FUNC(FUNCNAME)
+#undef FUNCNAME
+
+        /* Test 16: restore on fault for reg restored once (for app read)
+         * before crash.
+         */
+#define FUNCNAME test_asm_faultH
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+#ifdef X86
+        PUSH_CALLEE_SAVED_REGS()
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+
+        jmp      test16
+     test16:
+        mov      TEST_REG_ASM, DRREG_TEST_16_ASM
+        mov      TEST_REG_ASM, DRREG_TEST_16_ASM
+        nop
+        /* Read reg so that it is restored once. */
+        add      TEST_REG2_ASM, TEST_REG_ASM
+        mov      REG_XCX, 0
+        mov      REG_XCX, PTRSZ [REG_XCX] /* crash */
+        jmp      epilog16
+     epilog16:
+        add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
+        POP_CALLEE_SAVED_REGS()
+        ret
+#elif defined(ARM)
+        /* XXX i#3289: prologue missing */
+        b        test16
+     test16:
+        movw     TEST_REG_ASM, DRREG_TEST_16_ASM
+        movw     TEST_REG_ASM, DRREG_TEST_16_ASM
+        nop
+        /* Read reg so that it is restored once. */
+        add      TEST_REG2_ASM, TEST_REG_ASM, TEST_REG_ASM
+        mov      r0, HEX(0)
+        ldr      r0, PTRSZ [r0] /* crash */
+
+        b        epilog16
+    epilog16:
+        bx       lr
+#elif defined(AARCH64)
+        /* XXX i#3289: prologue missing */
+        b        test16
+     test16:
+        movz     TEST_REG_ASM, DRREG_TEST_16_ASM
+        movz     TEST_REG_ASM, DRREG_TEST_16_ASM
+        nop
+        /* Read reg so that it is restored once. */
+        add      TEST_REG2_ASM, TEST_REG_ASM, TEST_REG_ASM
+        mov      x0, HEX(0)
+        ldr      x0, PTRSZ [x0] /* crash */
+
+        b        epilog16
+    epilog16:
+        ret
+#endif
+        END_FUNC(FUNCNAME)
+#undef FUNCNAME
+
+        /* Test 17: restore on fault for gpr reserved in multiple phases
+         * with overlapping but not nested spill regions. In this case,
+         * the app value changes slots, from the one used in app2app
+         * phase, to the one used in insertion phase.
+         */
+#define FUNCNAME test_asm_faultI
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+#ifdef X86
+        PUSH_CALLEE_SAVED_REGS()
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+
+        jmp      test17
+     test17:
+        mov      TEST_REG_ASM, DRREG_TEST_17_ASM
+        mov      TEST_REG_ASM, DRREG_TEST_17_ASM
+        /* app2app phase will reserve TEST_REG2_ASM here. */
```

Comment:
Fixed the comment here locally; will update in next push. It should be `TEST_REG_ASM` instead of `TEST_REG2_ASM`.